plugin/trino-iceberg/src/main/java/io/trino/plugin/iceberg/UtcConstraintExtractor.java

```diff
@@ -284,12 +299,20 @@ private static Optional<Domain> unwrapDateTruncInComparison(String unit, Functio
     {
         Type type = constant.getType();
         checkArgument(constant.getValue() != null, "Unexpected constant: %s", constant);
-        checkArgument(type.equals(TIMESTAMP_TZ_MICROS), "Unexpected type: %s", type);
 
-        // Normalized to UTC because for comparisons the zone is irrelevant
-        ZonedDateTime dateTime = Instant.ofEpochMilli(((LongTimestampWithTimeZone) constant.getValue()).getEpochMillis())
-                .plusNanos(LongMath.divide(((LongTimestampWithTimeZone) constant.getValue()).getPicosOfMilli(), PICOSECONDS_PER_NANOSECOND, UNNECESSARY))
-                .atZone(UTC);
+        ZonedDateTime dateTime;
+        int precision = ((TimestampWithTimeZoneType) type).getPrecision();
+        if (precision <= MAX_SHORT_PRECISION) {
+            // Normalized to UTC because for comparisons the zone is irrelevant
+            dateTime = Instant.ofEpochMilli(unpackMillisUtc((long) constant.getValue()))
+                    .atZone(UTC);
+        }
+        else {
+            // Normalized to UTC because for comparisons the zone is irrelevant
+            dateTime = Instant.ofEpochMilli(((LongTimestampWithTimeZone) constant.getValue()).getEpochMillis())
+                    .plusNanos(LongMath.divide(((LongTimestampWithTimeZone) constant.getValue()).getPicosOfMilli(), PICOSECONDS_PER_NANOSECOND, UNNECESSARY))
```

Comment:
it's good to use UNNECESSARY here, but we should earlier assert that precision <= 9